src/verification/stream_split.lean

```diff
@@ -0,0 +1,117 @@
+import verification.stream_props
+
+variables {α ι : Type}
+
+section streams
+
+-- valid as long as the current index is in w
+@[simps]
+def substream (s : Stream ι α) (w : set ι) : Stream ι α :=
+{ σ := s.σ,
+  valid := λ p, ∃ h, s.index p h ∈ w,
+  ready := λ p, s.ready p,
+  next := λ p h, s.next p h.fst,
+  index := λ p h, s.index p h.fst,
+  value := λ p h, s.value p h,
+}
+
+variables {s : Stream ι α} {w : set ι}
+
+@[simp]
+private lemma substream.next'_eq {x : s.σ} :
+(substream s w).valid x → (substream s w).next' x = s.next' x :=
+λ h, by rw [Stream.next'_val h, substream_next, Stream.next'_val]
+
+private lemma substream.valid_subsumes {n : ℕ} {x : s.σ} :
+(substream s w).valid ((substream s w).next'^[n] x) → s.valid (s.next'^[n] x) :=
+begin
+  induction n with n ih generalizing x,
+  { simp only [function.iterate_zero, substream_valid],
+    exact Exists.fst },
+  { intro h,
+    have hxv := Stream.next'_valid' _ h,
+    rw [function.iterate_succ_apply] at h,
+    simpa [hxv] using ih h }
+end
+
+private lemma substream.bound_valid {B : ℕ} {x : s.σ} :
+s.bound_valid B x → ∀ w, (substream s w).bound_valid B x :=
+begin
+  simp_rw bound_valid_iff_next'_iterate,
+  induction B with n ih generalizing x,
+  { simp_rw [function.iterate_zero_apply, substream_valid, not_exists],
+    intros; contradiction },
+  { intros hnv w,
+    exact mt substream.valid_subsumes hnv }
+end
+
+end streams
+
+section stream_exec
+
+@[simps]
+def Stream.split (r : ι → ι → Prop) (s : StreamExec ι α) : Stream (quot r) (StreamExec ι α) :=
+{ σ := s.stream.σ × option (quot r), -- keep track of the last element seen
+  valid := λ p, s.stream.valid p.1,
+  -- TODO: instead of making the stream merely non-ready while skipping,
+  -- we should change next to recurse until it reaches the next window.
+  -- Will need to prove well-foundedness using bound_valid.
+  ready := λ p, s.stream.ready p.1 ∧
+                s.stream.bound_valid s.bound p.1 ∧ -- this is weird
```

Comment:
Having `bound_valid` be part of `ready` condition is a bit weird, but captures the fact that, at the `Stream` level, the state `p.1` is essentially an arbitrary state. (We know the state `p.1` is not _actually_ arbitrary since it has to be reachable from `s.state`, but nothing within `Stream` itself guarantees this.)

It seems easy enough to prove something like
```
s.bound_valid B x → s.bound_valid B (s.stream.next'^[n] x)
```
though.